libsodium-sys/build.rs

```diff
@@ -42,55 +94,406 @@ fn main() {
         .expect("Couldn't write bindings!");
 }
 
-
 /* Must be called when SODIUM_LIB_DIR or SODIUM_INC_DIR is set to any value
 This function will set `cargo` flags.
 Return: SODIUM_INC_DIR
 */
-fn find_libsodium_env() -> String {
+fn find_libsodium_env() -> Option<String> {
     let lib_dir = env::var("SODIUM_LIB_DIR")
         .expect("SODIUM_LIB_DIR must be set because SODIUM_INC_DIR is set. Error");
     let inc_dir = env::var("SODIUM_INC_DIR")
         .expect("SODIUM_INC_DIR must be set because SODIUM_LIB_DIR is set. Error");
 
-    let mode = match env::var_os("SODIUM_STATIC") {
-        Some(_) => "static",
-        None => "dylib",
-    };
-
-    if cfg!(target_env = "msvc") {
-        println!("cargo:rustc-link-lib={0}=libsodium", mode);
-    } else {
-        println!("cargo:rustc-link-lib={0}=sodium", mode);
-    }
     println!("cargo:rustc-link-search=native={}", lib_dir);
 
-    inc_dir
+    Some(inc_dir)
 }
 
 /* Must be called when no SODIUM_LIB_DIR and no SODIUM_INC_DIR env vars are set
 This function will set `cargo` flags.
 Return: the first include path from vcpkg
 */
 #[cfg(target_env = "msvc")]
-fn find_libsodium_pkg() -> String {
-    let lib = vcpkg::probe_package("libsodium")
-        .expect("Could not find libsodium on this system");
```

Comment:
I personally prefer a strict mode.

Reply:
@kpp - to clarify, are you suggesting a mode where if the lib is not found on the system, it errors rather than downloading / compiling? Something like `SODIUM_REQUIRE_SYSTEM`?

Reply:
Something like "Don't break the existing process". A user may want to see errors why the system library was not found especially when they set `SODIUM_LIB_DIR` and `SODIUM_INC_DIR`.

I suggest to build from src only when there are no env `SODIUM_LIB_DIR` and `SODIUM_INC_DIR` variables, and when there is `SODIUM_FROM_SRC` is set. We should report an error when there are `SODIUM_LIB_DIR` and `SODIUM_INC_DIR` and `SODIUM_FROM_SRC` are set.

Reply:
hmm. I think that makes sense for when `SODIUM_LIB_DIR` or `SODIUM_INC_DIR` are set, and can make that change.

What do you think about when there are no environment variables? My main hope is that this crate can be usable in a majority of environments without any special configuration... which probably means building from source, especially once https://github.com/sodiumoxide/sodiumoxide/issues/231 is included.

Reply:
If there are no env variables we should do everything to make `sodiumoxide` compile, first search for a system library then try to build it from source. The main question is whether we should build libsodium static in this case because if there is no system library installed the user will get a missing .so when they try to run binary.

Reply:
@kpp - updated per this conversation.
- if `SODUIM_LIB_DIR` or `SODIUM_INC_DIR` are specified, will not build from source, and will error if `SODIUM_FROM_SRC` is specified
- will build from source if nothing is specified but a system lib is not found